File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/Proxy/ManagedProxyListProvider.java
package ru.kmorozov.gbd.core.logic.Proxy;

import java.util.Optional;

public class ManagedProxyListProvider {

    private final IProxyListProvider parentProvider;
    private final int timeout;

    public ManagedProxyListProvider(final IProxyListProvider parentProvider, final int timeout) {
        this.parentProvider = parentProvider;
        this.timeout = timeout;
    }

    public ManagedProxyListProvider(final int timeout) {
        this(AbstractProxyListProvider.getInstance(), timeout);
    }

    public HttpHostExt getProxy() {
        if (this.checkReady(HttpHostExt.NO_PROXY))
            return HttpHostExt.NO_PROXY;

        if (!this.parentProvider.proxyListCompleted())
            this.parentProvider.processProxyList(UrlType.JSTOR);

        Optional<HttpHostExt> opProxy;
        do {
            try {
                Thread.sleep(100L);
            } catch (final InterruptedException e) {
                e.printStackTrace();
            }

            opProxy = this.parentProvider
                    .getParallelProxyStream()
                    .filter(this::checkReady)
                    .filter(HttpHostExt::isAvailable)
                    .findFirst();
        }
        while (!opProxy.isPresent());

        return opProxy.get();
    }

    private boolean checkReady(final HttpHostExt proxy) {
        return System.currentTimeMillis() - proxy.getLastUsedTimestamp() > (long) this.timeout;
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/proxy/providers/ManagedProxyListProvider.kt
package ru.kmorozov.gbd.core.logic.proxy.providers

import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.UrlType
import java.util.*

class ManagedProxyListProvider(private val parentProvider: IProxyListProvider, private val timeout: Int) {

    val proxy: HttpHostExt
        get() {
            if (checkReady(HttpHostExt.NO_PROXY))
                return HttpHostExt.NO_PROXY

            parentProvider.processProxyList(UrlType.JSTOR)

            var opProxy: Optional<HttpHostExt>
            do {
                try {
                    Thread.sleep(100L)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

                opProxy = parentProvider
                        .parallelProxyStream
                        .filter { this.checkReady(it) }
                        .filter { it.isAvailable }
                        .findFirst()
            } while (!opProxy.isPresent)

            return opProxy.get()
        }

    constructor(timeout: Int) : this(AbstractProxyListProvider.INSTANCE, timeout)

    private fun checkReady(proxy: HttpHostExt): Boolean {
        return System.currentTimeMillis() - proxy.lastUsedTimestamp > timeout.toLong()
    }
}
